src/nvim/option.c

```diff
@@ -1408,24 +1415,22 @@ static void do_set_option(int opt_flags, char **argp, bool *did_show, char *errb
     return;
   }
 
+  if (!(flags & P_BOOL) && vim_strchr("=:&<", nextchar) == NULL) {
+    *errmsg = e_invarg;
+    return;
+  }
+
   int value_checked = false;
-  if (flags & P_BOOL) {                       // boolean
+  if (flags & P_BOOL) {        // boolean
     do_set_bool(opt_idx, opt_flags, prefix, nextchar, afterchar, varp, errmsg);
-  } else {  // Numeric or string.
-    if (vim_strchr("=:&<", nextchar) == NULL || prefix != 1) {
```

Comment:
Why is this check moved to two separate places instead of one?

Reply:
To be clearer, I don't think this should be in `validate_opt_idx()`.

Reply:
> Why is this check split and moved to two separate places instead of one?

This is the wrong position for both checks, and the checks aren't really related.

> To be clearer, I don't think this should be in validate_opt_idx().

How come? It's similar to other checks in that function. Not all the checks require `opt_idx`. Some just check `flags` and `opt_flags` with other random bits of state.